datacompy/spark_core.py

```diff
@@ -0,0 +1,1023 @@
+
+    Attributes
+    ----------
+    spark_session : pyspark.sql.SparkSession
+        A ``SparkSession`` to be used to execute Spark commands in the comparison.
+    df1_unq_rows : pyspark.sql.DataFrame
+        All records that are only in df1 (based on a join on join_columns)
+    df2_unq_rows : pyspark.sql.DataFrame
+        All records that are only in df2 (based on a join on join_columns)
+    """
+
+    def __init__(
+        self,
+        spark_session,
+        df1,
+        df2,
+        join_columns=None,
+        abs_tol=0,
+        rel_tol=0,
+        df1_name="df1",
+        df2_name="df2",
+        ignore_spaces=False,
+        ignore_case=False,
+    ):
+
+        if not isinstance(df1, pyspark.sql.DataFrame):
+            raise TypeError("df1 must be a Spark DataFrame")
+        if not isinstance(df2, pyspark.sql.DataFrame):
+            raise TypeError("df2 must be a Spark DataFrame")
+
+        self.spark = spark_session
+        self._any_dupes = False
+
+        if join_columns is None:
+            raise Exception("Please provide join_columns")
+        elif isinstance(join_columns, (str, int, float)):
+            self.join_columns = [str(join_columns).lower()]
+        else:
+            self.join_columns = [str(col).lower() for col in join_columns]
+
+        self.df1 = df1.toDF(*[str(c).lower() for c in df1.columns])
+        self.df2 = df2.toDF(*[str(c).lower() for c in df2.columns])
+
+        self.df1_name = df1_name
+        self.df2_name = df2_name
+        self.abs_tol = abs_tol
+        self.rel_tol = rel_tol
+        self.df1_unq_rows = self.df2_unq_rows = self.intersect_rows = None
+        self.column_stats = []
+        self._compare(ignore_spaces, ignore_case)
+
+    @BaseCompare.df1.setter
+    def df1(self, df1):
+        """Check that it is a dataframe and has the join columns"""
+        self._df1 = df1
+        self._validate_dataframe("df1")
+
+    @BaseCompare.df2.setter
+    def df2(self, df2):
+        """Check that it is a dataframe and has the join columns"""
+        self._df2 = df2
+        self._validate_dataframe("df2")
+
+    def _validate_dataframe(self, index):
+        """Check that a dataframe has the join columns
+
+        Parameters
+        ----------
+        index : str
+            The "index" of the dataframe - df1 or df2.
+
+        Returns
+        -------
+        None
+        """
+        dataframe = getattr(self, index)
+
+        # Check if join_columns are present in the dataframe
+        if not set(self.join_columns).issubset(set(dataframe.columns)):
+            raise ValueError("{} must have all columns from join_columns".format(index))
+
+        if len(set(dataframe.columns)) < len(dataframe.columns):
+            raise ValueError("{} must have unique column names".format(index))
+
+        if (
+            dataframe.drop_duplicates(subset=self.join_columns).count()
+            < dataframe.count()
+        ):
+            self._any_dupes = True
```

Comment:
On the Spark side I think we have to be more mindful about the amount of time that actions on the dataframes will take, as the datasets may be arbitrarily large. This will walk through all rows of each dataframe, dedup them, and count both the deduplicated and original datasets. But if duplicates are found, it will perform the entire deduplication again in `_compare`. That could be an expensive operation and we should avoid trying to do it twice. That was the motivation behind [this code](https://github.com/capitalone/datacompy/blob/37b612a59b96e2ba41f1f1e7a08ac8972309dc16/datacompy/sparkcompare.py#L192-L199).